### PR TITLE
Pages functions bundle rework

### PR DESCRIPTION
**Pull request by Skye-31**
- [x] Building functions
- [x] Worker.js support
- [x] Build plugins
- [x] Make pages publish work
- [x] Some way of taking in D1 stuff from the CI to pass it to the shim, CC @WalshyDev 
- [x] Types from `@cloudflare/types`

### Cool side notes
- [x] Fixes the error "_worker.js is importing a file..." when it isn't importing one 🎉   

### Known issues
- ~~Running your own _routes.json with functions (not _worker.js) causes a crash~~

### CODE REVIEW

**Comment by WalshyDev** [comment]
For the passing of bindings, we need it to:
- Support all binding types without additional work inside our CI or wrangler
- Provide the info we need (name + type)
- Not be too jank (it'll definitely be a little jank but... that's ok... i guess...)

Initial proposal:
`--bindings "KV=kv_namespace,R2=r2_bucket,D1=d1_database"`
`--bindings <name=<type>,...`

This gives us all the info we need in an easy way and makes it simple for us to turn into a proper object. If anyone has any better ideas pls say

**Comment by Skye-31** [comment]
My concern is that some bindings may require different properties, such as IDs etc. It would almost be better for the CI to run the function from wrangler directly rather than running a CLI command. (like unstable_dev is used). This should be relatively simple to port and would avoid ugly string parsing?

**Comment by WalshyDev** [comment]
> It would almost be better for the CI to run the function from wrangler directly rather than running a CLI command. (like unstable_dev is used). This should be relatively simple to port and would avoid ugly string parsing?

use the wrangler API you mean? We'd still need to figure out what data we want to pass and pass it through. Dumping everything in an unstructured and unpredictable way isn't very feasible.

**Comment by Skye-31** [comment]
Yes, using the wrangler API is what I meant. 
Would it be plausible to give it the project as an argument? -- that way it could have identical logic to publish

File: packages/wrangler/src/bundle.ts [hunk 1]
```diff
@@ -235,24 +249,37 @@ export async function bundleWorker(
...
 			path.resolve(getBasePath(), "templates/experimental-local-cache-stubs.js")
 		);
 	}
 
+	const outFilepath = isOutfile
+		? path.isAbsolute(destination)
+			? destination
+			: entry.directory.endsWith("functions")
+			? path.join(entry.directory, "..", destination)
+			: path.join(entry.directory, destination)
```
Comment (WalshyDev) [inline packages/wrangler/src/bundle.ts:265]: Triple ternary monkaS, can we pls avoid that? Not very readable, some ifs or something would be nicer 🙃

File: packages/wrangler/src/dev/local.tsx [hunk 2]
```diff
@@ -497,6 +497,14 @@ export function setupMiniflareOptions({
 	miniflareCLIPath: string;
 	forkOptions: string[];
 } {
+	const defaultRules = DEFAULT_MODULE_RULES.slice();
```
Comment (WalshyDev) [inline packages/wrangler/src/dev/local.tsx:500]:
```suggestion
	const defaultRules = DEFAULT_MODULE_RULES;
```

File: packages/wrangler/src/pages/dev.tsx [hunk 3]
```diff
@@ -468,6 +474,7 @@ export const Handler = async ({
 			compatibilityDate,
 			compatibilityFlags,
 			nodeCompat,
+			bundle: false,
```
Comment (WalshyDev) [inline packages/wrangler/src/pages/dev.tsx:477]: Can we have a comment on why we make this false

File: packages/wrangler/src/dev/local.tsx (hunk 2 above)
Comment (WalshyDev) [reply packages/wrangler/src/dev/local.tsx:505]: ah, I see, yeah ok. I'd prefer we not have to do this but ok

Comment (Skye-31) [reply packages/wrangler/src/dev/local.tsx:505]: It was just a safety measure, I can remove if desired 🙂

**Comment by WalshyDev** [comment]
Settled on passing the bindings object through as a JSON. That way we always have support for infinite binding properties

so, it'll look like:
```
--bindings = '{"kv_namespaces": {"KV": { "namespace_id": "hash" } } }'
```

File: packages/wrangler/src/pages/build.tsx [hunk 4]
```diff
@@ -81,6 +81,18 @@ export function Options(yargs: Argv) {
...
+				describe: "Build for local development",
+				default: false,
+				type: "boolean",
+				hidden: true,
+			},
+			bindings: {
+				type: "string",
+				describe:
+					"Bindings used in Functions (used to register beta product shims)",
+				deprecated: true,
```
Comment (Skye-31) [inline packages/wrangler/src/pages/build.tsx:94]: Should this be `hidden` @GregBrimble?

Comment (GregBrimble) [reply packages/wrangler/src/pages/build.tsx:94]: Yeah, maybe not a bad shout to hide it as well. I'll add that.

File: packages/wrangler/src/pages/dev.tsx [hunk 5]
```diff
@@ -526,6 +526,7 @@ export const Handler = async ({
 			persist,
 			persistTo,
 			showInteractiveDevSession: undefined,
+			inspect: true,
```
Comment (Skye-31) [inline packages/wrangler/src/pages/dev.tsx:529]: This does not need to be explicitly set - it defaults to true. When set to true, it gives the "passing --inspect is not necessary message" when using pages dev, which is rather annoying

Comment (GregBrimble) [reply packages/wrangler/src/pages/dev.tsx:529]: Looks like that's defaulting to `false` to me

Comment (Skye-31) [reply packages/wrangler/src/pages/dev.tsx:529]: The options parameter which is destructured just after that overwrites it

Comment (GregBrimble) [reply packages/wrangler/src/pages/dev.tsx:529]: Ah, `inspect: undefined` works best: implicitly enables the inspector and means we don't get the message.

**Comment by GregBrimble** [comment]
Greg's to do list

- [ ] D1 bindings tests

**Comment by Skye-31** [comment]
With `bundle` enabled again - this entire file can be reverted. https://github.com/cloudflare/wrangler2/pull/2067/files#diff-dd2c9d6fc577a3da1b7cf0b6b4a0705a1c71eb5d99c372d5bdda99476362f975

**Review by GregBrimble — APPROVED** [review]

**Comment by GregBrimble** [comment]
LGTM. We're going to hold off merging until the next release, just so this can sit in `@beta` for a little while.

**Comment by GregBrimble** [comment]
Thanks @Skye-31 for doing this!!

**Pull request merged**